### PR TITLE
N21 419 create new tool modal

### PR DESCRIPTION
**Pull request by alweber-cap**
# Description

Added new Modal for CTL tools with expanded options

## Links to Tickets or other pull requests
https://github.com/hpi-schul-cloud/superhero-dashboard/pull/184
https://ticketsystem.dbildungscloud.de/browse/N21-239

## Changes
- Added new modal
- script changes to handle new fields

## Approval for review

- [ ] All points were discussed with the ticket creator, support-team or product owner. The code upholds all quality guidelines from the PR-template.

> Notice: Please remove the WIP label if the PR is ready to review, otherwise nobody will review it.

### CODE REVIEW

File: api.js [hunk 1]
```diff
@@ -10,7 +10,7 @@ const api = (req, { useCallback = false, json = true, version = 'v1' } = {}) =>
         headers['x-api-key'] = process.env.API_KEY;
     }
 
-    const baseUrl = process.env.BACKEND_URL || 'http://localhost:3030/api/';
+    const baseUrl = process.env.BACKEND_URL || 'http://localhost:3031/api/';
```
Comment (arnegns) [inline api.js:13]: Ups

File: controllers/ctltools.js [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+
+// secure routes
+router.use(authHelper.authChecker);
+
+router.get('/search', showTools);
+
+router.patch('/:id', getUpdateHandler('ltitools'));
+router.get('/:id', getDetailHandler('ltitools'));
+router.delete('/:id', getDeleteHandler('ltitools'));
+router.post('/', getCreateHandler('ltitools'));
```
Comment (arnegns) [inline controllers/ctltools.js:288]: I would like to delete all handler and routes which are not create/post.

File: controllers/tools.js [hunk 3]
```diff
@@ -11,11 +11,17 @@ const moment = require('moment');
...
 
-const getVersion = () => {
+const getHydraVersion = () => {
     return process.env.FEATURE_LEGACY_HYDRA_ENABLED ? 'v1' : 'v3';
 };
 
-const VERSION = getVersion();
+const getLtiVersion = () => {
+    return 'v1';
+    // return process.env.FEATURE_LEGACY_LTI_TOOLS_ENABLED ? 'v1' : 'v3';
```
Comment (arnegns) [inline controllers/tools.js:20]: Why you added this as comment?

File: controllers/tools.js [hunk 4]
```diff
@@ -231,31 +237,253 @@ const showTools = (req, res) => {
...
+    req.body.lti_version = req.body.lti_version || null;
+    req.body.lti_message_type = req.body.lti_message_type || null;
+    req.body.secret = (req.body.secret === PASSWORD ? undefined : req.body.secret);
+    req.body.isTemplate = true;
+    if(create || !req.body.isLocal) { // non-local (LTI) tools can be updated forever, local (OAuth2) only during creation
+        req.body.oAuthClientId = req.body.oAuthClientId || "";
+    } else {
+        req.body.oAuthClientId = undefined; // undefine the property prohibits database update
+    }
+    req.body.isLocal = (create ? !!req.body.isLocal : undefined);
```
Comment (arnegns) [inline controllers/tools.js:295]: Can we rename isLocal in new Tools or maybe delete it? When it is true we call hydra directly after tool creation or update.. i think this should be done in the server

File: controllers/tools.js [hunk 5]
```diff
@@ -231,31 +237,253 @@ const showTools = (req, res) => {
...
         });
     });
-});
+};
+
+
+// secure routes
+router.use(authHelper.authChecker);
+
+if(LTI_VERSION === 'v3') {
```
Comment (arnegns) [inline controllers/tools.js:449]: Can you rename this variable to something like USE_CTL or NEW_TOOLS or something like this. At the first moment i thought it has something to do with the version of lti tools

File: controllers/tools.js [hunk 6]
```diff
@@ -231,31 +237,253 @@ const showTools = (req, res) => {
...
+    router.get('/search', showToolsCtl);
+
+    router.patch('/:id', getUpdateHandlerCtl);
```
Comment (arnegns) [inline controllers/tools.js:452]: Please remove all handlers and routers which are not for the creation/post

**Review by arnegns — CHANGES_REQUESTED** [review]

File: controllers/tools.js (hunk 3 above)
Comment (alweber-cap) [reply controllers/tools.js:20]: removed

File: controllers/ctltools.js [hunk 7]
```diff
@@ -0,0 +1,302 @@
...
+            api(req, { version: HYDRA_VERSION }).get(`/oauth2/clients/${data.oAuthClientId}`).then(client => {
+              data.secret = PASSWORD;
+              data.redirect_url = client.redirect_uris.join(";");
+              data.token_endpoint_auth_method = client.token_endpoint_auth_method;
+              data.scope = client.scope;
+              data.frontchannel_logout_uri = client.frontchannel_logout_uri;
+              res.json(data);
+            });
+          } else {
+            data.secret = PASSWORD;
```
Comment (SevenWaysDP) [inline controllers/ctltools.js:137]: I can't understand why secret equals "******" in two places ?

**Review by SevenWaysDP — CHANGES_REQUESTED** [review]

File: controllers/ctltools.js (hunk 7 above)
Comment (MarvinOehlerkingCap) [reply controllers/ctltools.js:137]: the controllers/ctltools.js is mainly copied from the controllers/tools.js. We are doing necessary changes to it in https://github.com/hpi-schul-cloud/superhero-dashboard/pull/184

**Review by SevenWaysDP — APPROVED** [review]

**Pull request merged**